src/System.Diagnostics.Process/src/System/Diagnostics/Process.Unix.cs

```diff
@@ -291,15 +291,54 @@ private bool StartCore(ProcessStartInfo startInfo)
                 }
             }
 
-            int childPid, stdinFd, stdoutFd, stderrFd;
+            int stdinFd = -1, stdoutFd = -1, stderrFd = -1;
             string[] envp = CreateEnvp(startInfo);
             string cwd = !string.IsNullOrWhiteSpace(startInfo.WorkingDirectory) ? startInfo.WorkingDirectory : null;
 
+            bool setCredentials = !string.IsNullOrEmpty(startInfo.UserName);
+            uint userId = 0;
+            uint groupId = 0;
+            if (setCredentials)
+            {
+                (userId, groupId) = GetUserAndGroupIds(startInfo);
+            }
+
             if (startInfo.UseShellExecute)
             {
+                if (!IsValidVerb(startInfo.Verb))
+                {
+                    throw new Win32Exception(Interop.Errors.ERROR_NO_ASSOCIATION);
+                }
+
+                // On Windows, UseShellExecute of executables and scripts causes those files to be executed.
+                // To achieve this on Unix, we check if the file is executable (x-bit).
+                // Some files may have the x-bit set even when they are not executable. This happens for example
+                // when a Windows filesystem is mounted on Linux. To handle that, treat it as a regular file
+                // when exec returns ENOEXEC (file format cannot be executed).
+                bool isExecuting = false;
+                filename = ResolveExecutableForShellExecute(startInfo.FileName);
+                if (filename != null)
+                {
+                    argv = ParseArgv(startInfo);
+
+                    isExecuting = ForkAndExecProcess(filename, argv, envp, cwd,
+                        startInfo.RedirectStandardInput, startInfo.RedirectStandardOutput, startInfo.RedirectStandardError,
+                        setCredentials, userId, groupId,
+                        out stdinFd, out stdoutFd, out stderrFd,
+                        throwOnNoExec: false); // return false instead of throwing on ENOEXEC
+                }
+
                 // use default program to open file/url
-                filename = GetPathToOpenFile();
-                argv = ParseArgv(startInfo, filename);
+                if (!isExecuting)
+                {
+                    filename = GetPathToOpenFile();
+                    argv = ParseArgv(startInfo, filename);
+
+                    ForkAndExecProcess(filename, argv, envp, cwd,
+                        startInfo.RedirectStandardInput, startInfo.RedirectStandardOutput, startInfo.RedirectStandardError,
+                        setCredentials, userId, groupId,
+                        out stdinFd, out stdoutFd, out stderrFd);
+                }
```

Comment:
Indent above lines (313...340) one level deeper.
```suggestion
                    }
                } else {
                    throw new Win32Exception(Interop.Errors.ERROR_NO_ASSOCIATION);
                }
```

Organizing the Code this way, makes it easier to add codeblocks to implement other verbs like this (obviously before the `} else { throw ...` block):
```
                } else if (Verb.ToLower()=="runas"){
                    filename = ResolveExecutableForShellExecute("sudo");
                    argv = ParseArgv(startInfo, filename);
                    ForkAndExecProcess(filename, argv, envp, cwd,
                        startInfo.RedirectStandardInput, startInfo.RedirectStandardOutput, startInfo.RedirectStandardError,
                        setCredentials, userId, groupId,
                        out stdinFd, out stdoutFd, out stderrFd);
                }
```
<sub>(untested Code, don't use it without testing!)</sub>